staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -412,12 +417,23 @@ func (w *watchCache) SetOnEvent(onEvent func(*watchCacheEvent)) {
 
 func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]*watchCacheEvent, error) {
 	size := w.endIndex - w.startIndex
-	// if we have no watch events in our cache, the oldest one we can successfully deliver to a watcher
-	// is the *next* event we'll receive, which will be at least one greater than our current resourceVersion
 	oldest := w.resourceVersion + 1
```

Comment:
this might be easier to follow as a switch... the current version sets `oldest` four different ways at different nesting levels.

```go
var oldest uint64
switch {
// Once the watch event buffer is full, the oldest watch event we can deliver is the first one in the buffer.
case size >= w.capacity:
  oldest = w.cache[w.startIndex%w.capacity].resourceVersion

// If the watch event buffer isn't full, the oldest watch event we can deliver
// is one greater than the resource version of the last full list.
case w.listResourceVersion > 0:
  oldest = w.listResourceVersion + 1

// If we've never completed a list, use the resourceVersion of the oldest event in the buffer.
// This should only occur in tests that populate the buffer without performing a list/replace operation.
case size > 0:
  oldest = w.cache[w.startIndex%w.capacity].resourceVersion

default:
  return nil, fmt.Errorf(...)
}
```

Reply:
Good point. Done.